Framework for defining the maturity of OpenTelemetry and its SIGs (#232)
open-telemetry/oteps: text/0232-maturity-of-otel.md

# Framework for defining the maturity of OpenTelemetry and its SIGs

On 08 Mar 2023, the OpenTelemetry GC and TC held an OpenTelemetry Leadership summit, discussing various topics. One of the themes we discussed was establishing standard rules for describing the maturity of the OpenTelemetry project. This OTEP summarizes what was discussed there and is intended to have the wider community provide feedback.

This OTEP builds on what was previously communicated by the project, especially the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability).

The Collector's [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels) inspired the maturity levels.

## Motivation

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users.

Author: Users would have a reference on what features they can rely on. For instance, they should have a different maturity expectations between SDKs. While we, as a project, strive to get all languages following the same spec and having the same features, languages like .Net, Java, or Go will reach the "stable" level before others.

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users.

Reviewer: So, in summary, is the intended message for end-users that the project pays more attention/dedicates more resources to components owned by Tier-1 SIGs then?

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users.

Author: Not sure I would phrase it like that, but yes. I would say that this is what we are comfortable supporting, both from a roadmap criticality aspect (what you wrote) and from the healthiness of the community.

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users.

Author: I added an extra statement to clarify this.

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users.

Reviewer: Thanks! We can mark this as resolved

Only once we reach the "stable" level at the project level will we be ready to consider graduating OpenTelemetry at the CNCF, with all of the Tier-1 SIGs being in the context of the graduation, including their Tier-1 Components. This OTEP provides a framework for describing project maturity as a whole.

## Glossary

> * *SIG:* an official SIG within the OpenTelemetry project, like the Collector or Java. Some WGs are temporarily accepted, given they should be SIGs instead of WGs, like the Semantic Conventions WG.

Reviewer 2: > Some WGs are temporarily accepted, given they should be SIGs instead of WGs, like the Semantic Conventions WG.

Not sure what this sentence means.

Author: Sorry, I should have been clearer, and I accept suggestions. The wording is to account for the fact that we have WGs that should really be SIGs, as the "Semantic Conventions WG". So, when we say "SIG" in this doc, we mean actual SIGs plus those WGs that should be SIGs :-)

Even more context: during a TC (or GC?) call, someone asked what's the difference between a SIG and a WG. The consensus seems to be that WGs are geared towards and end-goal, and once that's reached, the WG is dissolved. SIGs are long-term, they don't have an end.

Author: I'm marking this as resolved, let me know if that's not the case.

Author: Actually, I'm reopening as I changed the text and would appreciate another review, verifying it's clearer now.

* *Tier-1 SIG:* a SIG within the OpenTelemetry project deemed critical by the GC/TC.
* *Candidate Tier-1 SIG:* a SIG that is deemed as critical for the future of the project, but might not be of the same maturity level yet. It's expected to reach that level eventually, and other Tier-1 SIGs should be prepared to accommodate the new component as needed.
* *Component:* module maintained by a SIG like a Collector exporter or an SDK sampler.
* *Tier-1 Component:* a component selected by project maintainers as critical for the SIG. For instance, the OTLP receiver will likely be selected as a Tier-1 Component of the Collector. See Tier-1 Components for more details.
* *Candidate Tier-1 Component:* similar to "Candidate Tier-1 SIG", but for a component within a SIG.
* *Dependency:* an OpenTelemetry SIG or component that is required by another SIG or component. For instance, the OTLP Logs Data Model from the Spec is required for the Collector's Logging support.  

## Explanation

> ### Maturity levels

Reviewer 3: @jpkrohling Can we add this clarification to the OTEP?

> ### Maturity levels

Author: Absolutely, I'm going to change this PR to include more details based on the questions and review comments.

> ### Maturity levels

Author: I added this clarification. Can you check the wording please?

> ### Maturity levels

Author: I see those as implementation details for the collector. While I think some of the questions are wider than just the collector (what to do with removed config options?), answering them should not be part of this OTEP but elsewhere, like the SIG Config.

If there's a specific question from those issues that you think we should clarify here, I'm happy to consider!

> ### Maturity levels

Reviewer: Looks good to me :+1: thanks for clarifying

> ### Maturity levels

Reviewer: I guess my concern is that wording like

> Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstance

seems too vague to be actionable. What does it mean to give prior notice? (e.g. for how many months?) What are special circumstances? (security issues?) 

I think these aspects are applicable to any software component in OpenTelemetry, not just the Collector's, but it may be the case that the specifics are different on each SIG, so I am happy to leave this out of scope if you think that is the case.

> ### Maturity levels

Author: I do have opinions on that, but I don't think this OTEP would be the right place to define this.

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Reviewer 2: > The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

I disagree with this. In java, we bundle many components of varying stability with the java agent. Waiting for all those components to be stable for the agent to be stable would mean that the agent is never stable. Instead, we have strategies like disabling certain less mature instrumentation by default.

Author: Good point. As a user of the Java agent, am I right to have the expectation that things will not break between versions when it comes to their configuration and telemetry. If I'm expected to deal with changed in metric names when migrating between versions, can I opt-out from those changes or should I just assume those metrics are not stable for me to consume it in alerts/dashboards?

Reviewer 2: I thin individual instrumentation modules will each have their own stability designations, and so user expectations will have to be on a case by case basis. I think the @open-telemetry/java-instrumentation-maintainers will have more thoughts about how to about stability guarantees of an agent comprised of many instrumentation modules.

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Author: I don't see a conflict, as the `otlpreceiver` itself can list which parts of it are considered tier-1. Should the OTLP receiver support a new signal tomorrow that is not a tier-1 signal yet, we'd certainly want to get that feature out while not marking it as "tier-1". Similarly, the Collector SIG can declare that only the config is part of the deliverable for the components, and not their internal APIs.

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Reviewer: My problem is not with the component having different stability levels for different signals, the problem I see is that the OTLP traces receiver is marked as stable while a dependency (the `config` package) is, (pressumably, since it's at `0.x`) not stable. 

So, one of the following must be true:
1. The current stability level for the OTLP traces receiver is wrong
2. The OTLP traces receiver does not "really" take a dependency on the `config` package
3. The rule *Only once all dependencies are marked as "stable" MAY a component be marked as "stable"* should not be followed

Which one is it?

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Author: IMO, it's the first option. It shouldn't have been marked stable if an underlying component isn't stable yet. What if the underlying config changes?

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Reviewer: See open-telemetry/opentelemetry-collector/issues/7964, let's continue the discussion there.

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Author: I'm marking this as resolved, as the discussion has been moved there.

> Only once all dependencies are marked as "stable" MAY a component be marked as "stable". Only once all Tier-1 Components of a SIG are "stable" MAY that SIG be marked as "stable".

Author: I'm marking this as resolved, please reopen if you think it's not.

#### Development

Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production.

#### Alpha

This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The configuration options might often change without backward compatibility guarantees.

#### Beta

> Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.

Reviewer 2: Can you give an example of a breaking change that doesn't involve the API, configuration, or generated telemetry?

Author: I can't think of one. Do you think I should not be explicitly listing those examples?

#### Deprecated

Development of this component is halted, and no further support will be provided. No new versions are planned, and the component might be removed from its included distributions. Note that new issues will likely not be worked on except for critical security issues. Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed.

#### Unmaintained

A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.

### Specification requirements

> The specification should declare all requirements for a specified component to be declared stable. When implementations are still missing features required by the specification, they can be at most Alpha status.

Reviewer 4: I think this would put .NET in Alpha, as it does not implement the Context spec and also has some other incompatibilities, e.g. https://github.com/open-telemetry/opentelemetry-dotnet/pull/3841. As far as I understand, this is due to it not providing a full OTel implementation but cooperating with the .NET runtime for providing required features. Probably a special exception is needed here.

Reviewer 3: cc @open-telemetry/dotnet-approvers

Reviewer 5: Agreed, an alpha designation does not make sense for .NET, but we are a unique API/SDK in some ways.

I haven’t taken a pass over the spec compliance matrix in some time, and it's possible it may need some updates.

I take some time next week to take a look at the matrix for .NET, submit a PR with updates as necessary, and review it with an eye for what our suggestion should be regarding this OTEP.

My thought off hand is that perhaps we should suggest a way to accurately reflect in the matrix deviations w.r.t. spec compliance that we do not consider affecting the SDK's maturity.

Author: We could have a provision here for exceptions, but I'm curious as to the message we send when we state that "Context" is required and .NET doesn't implement it.

Reviewer 4: I would for sure prefer it .NET implemented Context... But not sure if it is realistic. Or we could give .NET a N/A designation?

Author: Based on the Slack conversation, my understanding is that the parts of the spec that are not implemented by dotnet are optional only. No required aspects are left unimplemented.

Reviewer 5: I've responded in the Slack thread. Summarizing here as well.

With regards to the Context spec, I think .NET's [`RuntimeContext`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/RuntimeContext.cs) covers some of the concerns. There are other parts of the spec related to context that .NET has not implemented, e.g., span processor OnStart does not have an explicit context parameter. Though, RuntimeContext may be sufficient for many use cases.

Reviewer 4: I'm unresolving this because I don't think RuntimeContext covers any of the concerns as long as it's not used anywhere at all by the SDK and API, not even in the places it's required 😃

Reviewer 5: > I'm just saying that it technically does not fit the current stability definition.

Agreed. You have noted a couple examples where .NET does not comply with the spec, but I think the general point you raise is valid:

1. How do we evaluate stability/maturity?
2. Given that .NET maintainers have declared the API/SDK stable despite some deviations from the spec, is the maintainers' judgement sufficient to declare stability?
3. If it Is it possible for an API/SDK to be declared stable despite deviations from the spec, should we have a means to denote/identify deviations in the spec compliance matrix? If so, it would seem that the `+/-` designations in the matrix is insufficient. `-` seems to suggest that something just hasn't been implemented yet. `+` should probably mean it is 100% compliant with the spec. For items not 100% compliance, maybe we need to denote it either `N/A` (not going to implement) or `+*` (implemented but deviates w/ info captured in a footnote).

Author: Thank you all for the clarifications. I'll split this discussion into its own issue, especially after today's decision to get this OTEP to focus on the maturity level framework, leaving the concept of tier-1 and the list of tier-1 SIGs out for now.

Author: I was confident I had, but I can't find it. I just opened one, to handle the generic case: https://github.com/open-telemetry/opentelemetry-specification/issues/3673

### Tier-1 SIGs

This OTEP recommends the following SIGs to be included as part of the Tier-1 SIGs for OpenTelemetry:

* Spec General
* Semantic Conventions (WG)

> * Java

Reviewer 6: ```suggestion
* Java
* Collector
```

The Collector should also be Tier-1, no? 🙂

Reviewer 3: I can't remember why Python is not part of this group, but I'd be interested in adding it as well.

Author: I'll add this to the next GC/TC meeting, perhaps we can finalize a list there.

Author: Added to the agenda for July 13th, please record your vote!

> * Java auto-instrumentation

Reviewer 2: Java and Java auto-instrumentation are one SIG

Author: True, not sure where I saw two separate entries.

* Go

> * .Net

Reviewer 7: nit

```suggestion
* .NET
```

* JavaScript

Once this OTEP is approved, the TC and GC members will vote on the individual SIGs from the list above and are free to make further suggestions. A final list of Tier-1 SIGs shall be in the "community" repository.

### Tier-1 Components

> When discussing the components' maturity, we often debate what should be included in the evaluation criteria. To help clarify that, we introduce the concept of "Tier-1 SIGs" and "Tier-1 Components". Tier-1 Components are critical for a specific SIG: only when all Tier-1 Components are stable can a project be evaluated as stable. For instance, the "OTLP receiver" for the OpenTelemetry Collector could potentially be a "Tier-1 Component". Only after this component is declared stable may the Collector be declared stable, provided all other Tier-1 Components are also stable.

Reviewer: > Only after this component is declared stable may the Collector be declared stable, provided all other Tier-1 Components are also stable.

Assuming Tier-1 components are also considered dependencies, this requirement is a partial version of a previous requirement noted above,

> Only once all dependencies are marked as "stable" MAY a component be marked as "stable"

So either the Tier-1 requirement noted here is redundant or something else is wrong :)

Author: It is redundant. I was just being extra clear that the same requirements apply downstream.

Reviewer: Alright! You can mark this as solved then, just found it confusing

Reviewer: Clearer now indeed, thanks!

A component can be declared "Tier-1" only if they have the following characteristics:

* The component has at least the same maturity level as the SIG
* The component has a clear code owner
* The code owner is active within the project
* The maintainers of the project are comfortable supporting the component should the code owner decide to step down or leave the project

### Tier-1 Signals

Similar to a Tier-1 Component, a Tier-1 Signal must have stable support in order to declare OpenTelemetry as stable. Declaring a signal as Tier-1 means that all projects must also see that signal as required for declaring the project stable.

Not all signals will have the same requirements. For instance, Logs might not provide an instrumentation API specification.

List of Tier-1 signals and their capabilities:

* Logs
  * Data model
  * SDK specification
  * Instrumentation SDKs
  * Bridge API specification
* Metrics
  * Data model
  * API specification
  * Instrumentation APIs
  * SDK specification
  * Instrumentation SDKs
* Traces

>   * Data model

Reviewer 2: If only we had a trace data mode 😢

Author: Your comment intrigued me. Indeed, we don't seem to have a "data model" page under our Tracing spec, but what I would expect to see there is available as part of the API, like this: https://opentelemetry.io/docs/specs/otel/trace/api/#span

What I had in mind though was this: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto

In any case, do you believe there's something we should address in this PR here?

Reviewer 2: No, probably nothing to address in this PR.

  * API specification
  * Instrumentation APIs
  * SDK specification
  * Instrumentation SDKs

The lifecycle of a signal is defined in the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability). While the stability and versioning of the signals may relate to this OTEP, the previously linked document specifies the details on how to reach the "stable" level mentioned in this OTEP.

The "experimental" level from the signal versioning and stability document relates to the "alpha" and "beta" levels from this OTEP, "stable" and "deprecated" are equivalent to their counterparts on this document, and "removed" is non-existent here.

When a new signal reaches the "stable" level, OpenTelemetry TC and GC will ultimately decide whether to include it as part of OpenTelemetry's priority components, using the process described in [Promoting SIGs and components](#promoting-sigs-and-components).

### Promoting SIGs and components

> SIGs can request to be included in the Tier-1 List by contacting a TC or GC member. The TC/GC member will bring the request to the TC/GC via Slack or Zoom call, who will vote on the matter. A simple majority suffices for including the SIG. The SIG needs to have at least equal maturity as the project. For instance, if OpenTelemetry's maturity is set to Beta, the coming SIG has to be at least Beta as well, so that it doesn't downgrade the main project's maturity.

Reviewer 2: > A simple majority suffices for including the SIG.

Typo?

Reviewer 2: Not sure what I meant by this. Maybe was reading too fast originally? Anyway, closing.

# Framework for defining the maturity of OpenTelemetry and its SIGs

On 08 Mar 2023, the OpenTelemetry GC and TC held an OpenTelemetry Leadership summit, discussing various topics. One of the themes we discussed was establishing standard rules for describing the maturity of the OpenTelemetry project. This OTEP summarizes what was discussed there and is intended to have the wider community provide feedback.

This OTEP builds on what was previously communicated by the project, especially the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability).

The Collector's [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels) inspired the maturity levels.

## Motivation

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the project's maturity. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users. The public message we want to convey by stating that a SIG is a tier-1 SIG is that the SIG has both a healthy community and is seen as critical to the project as a whole.

Reviewer 8: > the community is faced with the question of the quality and maturity expectations of its diverse **set of components**

(emphasis mine) +1 that this is a problem. But then you're making a leap from this to the question of "SIG maturity", which is a different problem altogether. Why is SIG maturity a problem?

Author: "SIG maturity" is quite judgemental, and I think it's affecting what is really meant in the OTEP. Perhaps we don't need to have SIGs being "stable", but having SIGs report their tier-1 components and their stability levels? OTel would then be ready to be called stable when all components for all tier-1 SIGs are marked as stable. Would it help to remove this intermediary level?

Reviewer 9: SiG != github repository or contributor ecosystem necessarily.  IF we were to have any "bar" around stability and participation, I think it needs to be around things like release-velocity, major bug resolution time and contributor-health vs. attending the public meetings.

> Only once we reach the "stable" level at the project level will we be ready to consider graduating OpenTelemetry at the CNCF, with all of the Tier-1 SIGs being in the context of the graduation, including their Tier-1 Components. This OTEP provides a framework for describing project maturity as a whole.

Author: No, they are out of the scope for graduation. For instance, they will not be part of the security audit. If there's interest in having a security audit for a non-tier-1 SIG/Component, the SIG Security can still try to help them.

## Glossary

> * *SIG:* an official SIG within the OpenTelemetry project, like the Collector or Java. Given that some Working Groups (WGs) are effectively SIGs, like the Semantic Conventions WG, WGs will be accepted on an individual basis.

Reviewer 9: I'm not convinced Semantic Conventions WG (as the charter implies) is a SiG.  It could be we need to evolve into one, but that would be a change of charter.  Instead, we're attempting to grow individual ownership of sub-WG.

I think I understand the goals of stability here, but I don't want to implicitly be changing the charter of a working group.  Specifically the current SemConv WG is meant to finish process + tooling, at which point it can re-evaluate its charter and existence.

Author: That's great feedback, I did think the Semantic Conventions WG was not properly classified and should be a SIG. Other folks felt the same, but it's good to see that it's not the case.

Do you think I should leave the semantic conventions out of the otep for now?

* *Tier-1 SIG:* a SIG within the OpenTelemetry project deemed critical by the GC/TC.

> * *Candidate Tier-1 SIG:* a SIG that is deemed as critical for the future of the project, but might not be of the same maturity level yet. It's expected to reach that level eventually, and other Tier-1 SIGs should be prepared to accommodate the new component as needed.

Author: Correct. The idea is to have a clear list of things that will become Tier-1 in the future, so that other SIGs can prepare for this new requirement. For instance, if we define a hypothetical new signal for stacktraces, we might want all SDKs to support this before marking "stacktraces" as tier-1.

* *Component:* module maintained by a SIG like a Collector exporter or an SDK sampler.
* *Tier-1 Component:* a component selected by project maintainers as critical for the SIG. For instance, the OTLP receiver will likely be selected as a Tier-1 Component of the Collector. See Tier-1 Components for more details.
* *Candidate Tier-1 Component:* similar to "Candidate Tier-1 SIG", but for a component within a SIG.
* *Dependency:* an OpenTelemetry SIG or component that is required by another SIG or component. For instance, the OTLP Logs Data Model from the Spec is required for the Collector's Logging support.  

## Explanation

### Maturity levels

> SIGs and components of a project MUST have a declared maturity level established by SIG maintainers (SIGs). The maturity of the SIG is assessed by its maintainers, while the maturity of the components might be delegated by the maintainers to code owners. The maturity level for a SIG or component is, at most, the lowest level of its dependencies or priority components. For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

Reviewer 10: >For instance, if the Collector SIG maintainers declare the "otlpreceiver" component as a tier-1 component of the Collector core distribution, and the "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta".

I am not sure I understand the purpose of this rule. Can you please clarify why? I think it is possible that we use a dependency that is immature but hidden/encapsulated in a way that does not expose its immaturity. Do you still want the immaturity to bubble up anyway? Is your suggestion limited to publicly visible components only?

Author: This rule was derived by the Collector's rules. If otlpreceiver includes confighttp (alpha), it means that confighttp may change the way an endpoint is configured. Users of the otlpreceiver will be affected by changes to confighttp.

> Is your suggestion limited to publicly visible components only?

This is a good question. I have nothing against limiting the rule to publicly visible components, such as components that would affect the configuration, telemetry, or final behavior of the component. How does that sound?

> Only once all dependencies are marked as "stable" MAY a component be marked as "stable". Only once all Tier-1 Components of a SIG are "stable" MAY that SIG be marked as "stable".

Reviewer 9: A SiG is a group of people working a component, a component *itself* is the software delivered by those people.

I'm not sure we should conflate the two.  e.g. I'm fine if semantic conventions are considered a critical / tier-1 component, but I'm not sure the *working group* is.

Maybe it would make sense to focus this proposal on components?   Rather than talking about SiGs, we can focus on OWNERS/MAINTAINERS of those components.

Author: The idea is that SIG maintainers would have the power to determine which deliverables are included in the SIG's tier-1 list. For instance, the Collector might include the "core" distribution as "tier-1" but not the "contrib" distribution.

Author: In the context of this OTEP, we have a list of tier-1 SIGs, each one delivering components of different stability levels. Once all components for a given SIG are stable, that SIG is declared stable. Once all tier-1 SIGs are stable, can we call OpenTelemetry stable.

Reading that paragraph without this context makes it sound weird: there's no attempt to mark SIGs stable or not, it's just a way to see how far we are into the overall stability of the project, not a judgment of the SIG itself.

#### Unmaintained

A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.

### Specification requirements

The specification should declare all requirements for a specified component to be declared stable. When implementations are still missing features required by the specification, they can be at most Alpha status.

> ### Tier-1 SIGs

Reviewer 9: What's the criteria for a Tier-1 SiG vs. a Tier-1 Components?

For component, I think I can suggest a litmus test. For a SiG (group of people) I'm not sure I have an answer.

For example, Looking at this list, I'd argue we should target the following:

- Specification (all components marked stable)
- Semantic Conventions (similarly to the above)
- OTLP/Protocol (components marked stable)
- Collector-core distribution
- Collector-contrib distribution
- SDK distributions (Java, .NET, Go, JavaScript, Python)
- Instrumentation components deemed critical by Language SiGs (e.g. Java auto-agent)

Effectively, we ask ourselves:

- What are the bare minimum components needed for successful opentelemetry adoption?
- What pieces are needed for ongoing success of OpenTelemetry users?

Things like OpAMP we can exclude for now as the maturity and need rises.

Author: > For example, Looking at this list, I'd argue we should target the following:

From that list, I would argue that the contrib distribution should _not_ be part of it, as it includes components that are unlikely to be supported. This is why I think SIGs should list what they believe to be the right tier-1 deliverables. 

What I like about fixing the list of components here is that we know right now the list of components we'll aim for stabilizing before anything else.

Reviewer 9: This is why I think the overall list should be COMPONENTS, not SiGs.   I think it's possible if we classify SiGs we'll get some unfairness/bias against smaller SiGs that *are* ready to participate in stability of OTEL.

I'd rather see an approach where we do something like the following:

1. For each "component" in OTEL, it can only declare itself stable if it implements a *stable* release of all *stable* sections of the specification.
2. For utlity components like the collector, we require a strict separate of stable from unstable.   For example, I actually don't think we can consider the metrics portion of the Collector "done" until Prometheus support is stable.  While there is no specification to control the collector, I think we need some form of document denoting what we consider *key* or *minimum viable* functionality that collector must have to be considered stable.   Similarly, we need such a definition from the OTEL Operator, e.g.
3. We work w/ SiGs / Maintainers to understand which are able to comply with stability.   Areas of OTEL that are not participating will need to be clearly denoted, possibly via some form of convention.   E.g. if opentelemetry-rust *is* ready to participate in stability, it can do so simply by having stable components released.  If e.g. Haskell (https://github.com/ethercrow/opentelemetry-haskell) wants to be part of OTEL, it would be held to the same standards and can participate in the "stable" portion of otel.  _Note:  I think it's ok for us to put requirements on active community size._

Right now though, with this proposal, I'm not sure I could figure out what the criteria or value-function for deciding "tier-1 SiGs" would be.

Author: > I think we need some form of document denoting what we consider key or minimum viable functionality that collector must have to be considered stable.

Who's "we" here? I take it that you mean the TC, while this OTEP suggests it's up for the Collector maintainers to decide. 

> it can do so simply by having stable components released

One problem I saw with that is that Rust (sorry) could have a stable release there, but if one specific maintainer leaves the project, it leaves the project in a state where we cannot call stable anymore. That's why I think the willingness of a language/SIG has to count, but the ultimate call is on the TC.

> what the criteria or value-function for deciding "tier-1 SiGs" would be

The "Promoting SIGs and components" outlines the process but purposely not the criteria. It's left to the TC members to decide based on available information. If you think a set of criteria should be define beforehand, we could certainly come up with some, although I would really prefer not to have this list until we have gone through this a couple of times.

This all said, I'm leaning more towards listing components here in this OTEP and not SIGs. I think classifying SIGs as tier-1 has caused more harm than good.

This OTEP recommends the following SIGs to be included as part of the Tier-1 SIGs for OpenTelemetry:

* Spec General
* Semantic Conventions (WG)
* Collector
* Java
* Go
* .NET
* JavaScript

> Once this OTEP is approved, the TC and GC members will vote on the individual SIGs from the list above and are free to make further suggestions. A final list of Tier-1 SIGs shall be in the "community" repository.

Reviewer 11: What's the process for maintenance of this list?  Can SIGs be added? Removed?  At what cadence?  What are the requirements of the TC/GC vote?  If the answer is "a process to be developed by the TC/GC" that should probably be noted here.

Author: This is defined later in the document, under the sections "Promoting SIGs and components" and "Downgrading SIGs and components". I'm marking this comment as resolved, but let me know if there's something not covered there.

### Tier-1 Components

When discussing the components' maturity, we often debate what should be included in the evaluation criteria. To help clarify that, we introduce the concept of "Tier-1 SIGs" and "Tier-1 Components". Tier-1 Components are critical for a specific SIG: only when all Tier-1 Components are stable can a project be evaluated as stable. Example:

> * The "OTLP receiver" for the OpenTelemetry Collector could potentially be a "Tier-1 Component". Only after this component is declared stable may the Collector be declared stable, provided all other Tier-1 Components are also stable.

Reviewer 8: I do not understand the motivation here. Why do we need these derivation rules? If, as an end user, I am deploying a collector which includes `otlpreceiver` as `stable` and `config` module as `beta`, what value do I get from any maturity designation at the whole collector level? I think I already have all the info I need at the component level.

Author: You are right: as an end-user, I would care only about the collector's calculated stability. I can then assume that all components within the collector are of the same stability or higher.

However, as a maintainer of the project, I care about the stability of the individual components, so that I can make a decision on which components to include in the deliverable.

As a code owner for a component, I care about the stability of my component so that it can be included in a tier-1 deliverable.

Reviewer 8: > You are right: as an end-user, I would care only about the collector's calculated stability.

My point was exactly the opposite: as an end user I *do not* care about cumulative/calculated stability level, because it's largely useless being the lowest common denominator.

Reviewer 10: As a Collector user I don't care about the least stable Collector component if I am not going to use that component. It has no impact on me because by design components that I don't use are not even instantiated at runtime. On the other hand I do care about Collector's core modules of the codebase (e.g. pdata, etc) because they impact every single component, they are foundational dependency of every other component.

So, really it is not simply a minimum stability of all components included, there may be more user-specific nuance and situational nuance depending on what the dependency graph looks like. The calculated "overall" stability may or may not be the minimum of all components. We should probably leave out the calculation rule from the OTEP and let maintainers specify how to assess the "overall stability" if we think it is important to be documented.

Reviewer 8: > When I see that the Core distribution is stable, I assume I can use everything in it and expect it to work. Having a "the Collector is stable, except for the following components..." is a confusing message.

That's why we may want to avoid making any claims about stability of the Collector as a whole, since it is inherently meaningless. Linux has a whole bunch of subsystems of different maturities, does it have a notion of a single, distro-level "Stable"?

Author: Not sure about the Kernel (although I remember a time when a pair version was a stable one, and an odd version was experimental), but I consume the kernel as part of a distribution. When I install Fedora or Red Hat Enterprise Linux, I expect it to work out of the box, no matter what the individual versions are. LTS versions of different distributions are quite popular among folks running production-grade workloads because of that: they know that everything in the repositories for that distribution is supported.

Projects MUST list their Tier-1 Components in the main readme file. A Tier-1 Component MAY be hosted under the project's "contrib" repository. Not all components under the main repository (sometimes called the "core" repository) are automatically Tier-1 Components.

A component can be declared "Tier-1" only if they have the following characteristics:

* The component has at least the same maturity level as the SIG
* The component has a clear code owner
* The code owner is active within the project
* The maintainers of the project are comfortable supporting the component should the code owner decide to step down or leave the project

> ### Tier-1 Signals

Reviewer 9: I feel like this should just be calling out which pieces of the specification we consider "tier-1" or core, vs. going signal-by-signal.

E.g. Exemplars, Resources, Context are all *critical* to the project and aren't really called out in this model.

Author: The spec as a whole should be covered as part of the Spec SIG, included in the Tier-1 SIGs. I felt like we should list the signals here, but perhaps we should just remove this entire section? Stability for each signal is defined within the Spec SIG and it could list the individual signals (and other parts, like the ones you mentioned) in the tier-1 components?

Author: Given how the OTEP is right now, I think this doesn't belong here anymore, but this might still be important. The Collector part should be covered by this: https://github.com/open-telemetry/opentelemetry-collector/issues/8555 . What else do you think we need in the same area?

Author: @jsuereth, is there something to clarify around this, or can this be marked as resolved?

Author: I'm marking this as resolved, let me know if there's still something to address here.

Not all signals will have the same requirements. For instance, Logs might not provide an instrumentation API specification.

List of Tier-1 signals and their capabilities:

* Logs
  * Data model
  * SDK specification
  * Instrumentation SDKs
  * Bridge API specification
* Metrics
  * Data model
  * API specification
  * Instrumentation APIs
  * SDK specification
  * Instrumentation SDKs
* Traces
  * Data model
  * API specification
  * Instrumentation APIs
  * SDK specification
  * Instrumentation SDKs

The lifecycle of a signal is defined in the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability). While the stability and versioning of the signals may relate to this OTEP, the previously linked document specifies the details on how to reach the "stable" level mentioned in this OTEP.

The "experimental" level from the signal versioning and stability document relates to the "alpha" and "beta" levels from this OTEP, "stable" and "deprecated" are equivalent to their counterparts on this document, and "removed" is non-existent here.

> When a new signal reaches the "stable" level, OpenTelemetry TC and GC will ultimately decide whether to include it as part of OpenTelemetry's priority components, using the process described in [Promoting SIGs and components](#promoting-sigs-and-components).

Reviewer 11: ```suggestion
When a new signal reaches the "stable" level, OpenTelemetry TC and GC will ultimately decide whether to include it as part of OpenTelemetry's Tier-1 components, using the process described in [Promoting SIGs and components](#promoting-sigs-and-components).
```

Author: I guess you found what was the term we had before tier-1 :-) Thank you for spotting this!

### Promoting SIGs and components

SIGs can request to be included in the Tier-1 List by contacting a TC or GC member. The TC/GC member will bring the request to the TC/GC via Slack or Zoom call, who will vote on the matter. A simple majority suffices for including the SIG. The SIG needs to have at least equal maturity as the project. For instance, if OpenTelemetry's maturity is set to Beta, the coming SIG has to be at least Beta as well, so that it doesn't downgrade the main project's maturity.

> Components can request project maintainers to be included in the Tier-1 list via a GitHub issue created on the main repository of the SIG. The SIG maintainers will then vote on the matter. A simple majority suffices for including the component. The component needs to have at least equal maturity as the SIG so that it doesn't downgrade the SIG's maturity. If SIG maintainers have strong reasons to include a component with a lower status, they can request an exception with the GC/TC. A supermajority vote is required to accept the request until the project reaches the "Stable" level so the project's scope doesn't continuously grow. After we reach the "Stable" level, a simple majority decision will be sufficient.

Reviewer 11: >If SIG maintainers have strong reasons to include a component with a lower status, they can request an exception with the GC/TC. A supermajority vote is required to accept the request until the project reaches the "Stable" level so the project's scope doesn't continuously grow. After we reach the "Stable" level, a simple majority decision will be sufficient.

This feels wrong to me.  It seems like it should be the other way around and that it should become significantly more difficult to add a component with a lower stability level once the `Stable` level is reached.  Always requiring a supermajority would also make sense to me.

Is "supermajority" defined somewhere in the governing documents that can be referenced here?  That term could have a wide variety of interpretations.

Author: > It seems like it should be the other way around and that it should become significantly more difficult to add a component with a lower stability level once the Stable level is reached

Both would require a vote at the GC/TC level, so it's already hard. I think we had simple majority for both at first, but to avoid feature creep (to use @dyladan's term), we changed so that an approval by 2/3s of the GC/TC members would be required before we get into a stable state.

I'm good either way: making it a simple or super majority for all cases.

@dyladan, what's your opinion on this?

Author: About defining supermajority: we don't have this definition anywhere, I'm afraid. I thought this was a well-understood concept, but Wikipedia lists three possible values: 66%, 60%, and 55%. I meant 66% (2/3s) here, I'll add a definition somewhere in this doc.

Reviewer 12: I also thought it was a well understood concept but I guess not. @jpkrohling is right this was originally my idea and it was to prevent too many new things being added that would delay graduation. Maybe we can keep it as a simple majority of both GC and TC (separate votes as they represent separate interests in their decision making). I think everyone involved there is aware of the risk of feature creep delaying graduation.

Author: I changed this to be a simple majority.

This OTEP builds on what was previously communicated by the project, especially the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability).

The Collector's [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels) inspired the maturity levels.

## Motivation

Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the maturity for components delivered by SIGs in the name of the project. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users using a unified nomenclature.

## Explanation

### Maturity levels

> Deliverables of a SIG MUST have a declared maturity level, established by SIG maintainers (SIGs), likely with the input of the code owners. The maturity level of a deliverable is, at most, the lowest level of the user-visible dependencies. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" can be at most "beta". The OpenTelemetry Collector "core" distribution is also affected so that it can be at most "beta". Only once all dependencies are marked as "stable" MAY a component be marked as "stable".

Reviewer 10: >The maturity level of a deliverable is, at most, the lowest level of the user-visible dependencies.

I am not entirely sure about this part. Looks for example at Kubernetes. They have individual features marked [alpha] ([for example](https://kubernetes.io/docs/concepts/containers/images/#maximum-parallel-image-pulls)), however the entire Kubernetes is not marked [alpha], it has a 
stable release 1.x.

I think this is the right approach. Mark individual components appropriately and if the entire thing makes sense for production use then it can still be marked as stable even if there are less mature components in it.

I would prefer to leave it to individual SIGs to decide how they derive the maturity of the whole thing from its constituents.

Reviewer 8: I agree with @tigrannajaryan. I don't like this formulaic approach which would result in every top level components being "beta".

Author: There were similar concerns around the Java Agent and I agree with that. I rephrased this.

#### Development

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer 10: >User feedback is desired, especially regarding the user experience

Do we want this at this this level? I wonder if that may be premature. Should we instead start recommending this from "Alpha" level?

Author: As a developer of a component, I would appreciate to hear about possible use-cases for my component. I wouldn't care too much about performance issues, though.

Author: @tigrannajaryan, do you think my points are reasonable, or do you think the text needs a change?

Reviewer 10: I prefer to avoid actively soliciting feedback for a "Development" component that is just introduced. If I have just created it, and it likely is half-broken (as expected for this level) I don't want to be receiving comments from users saying it is broken. I know it already. I understand we are seeking feedback because it may be valuable, but I suspect signal-to-noise ratio is going to be low. 

I am not strongly opposed to keeping the text as is if that's what the majority of the community prefers, but would like to see a few voices in favour of it.

Author: How about rewording to something like this:

> User feedback around the UX of the component is desired, such as for configuration options, component observability, technical implementation details, and planned use-cases for the component.

Reviewer 10: Sounds good.

#### Alpha

> This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The configuration options might often change without backward compatibility guarantees.

Reviewer 10: ```suggestion
This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The component's interface and configuration options might often change without backward compatibility guarantees.
```

#### Beta

Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.

> 

Reviewer 10: As per [this comment](https://github.com/open-telemetry/opentelemetry-specification/issues/3459#issuecomment-1648448391) "Frozen" in Go is a different concept. I am not sure it needs to be part of maturity levels.

#### Stable

The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

> #### Deprecated

Reviewer 10: Is "unmaintained" a necessary intermediary step before moving to "deprecated"?

Author: @tigrannajaryan, can I mark this one as resolved, or do you think this needs to be changed?

Reviewer 13: Typically, components may decide to move to deprecation directly through a voluntary change taken by codeowners with approval of the maintainers. This has happened multiple times in the collector contrib project.

#### Unmaintained

A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.

## Trade-offs and mitigations

This OTEP allows SIG maintainers to declare the maturity of the SIG's deliverables without declaring which ones are key for OpenTelemetry. When, and if, this is needed, a new OTEP may be created using the maturity levels as a possible framework.

## Prior art and alternatives

> * The specification status has a ["Component Lifecycle"](https://opentelemetry.io/docs/specs/status/) description, with definitions that might overlap with some of the levels listed in this OTEP.

Reviewer 14: it'd be great to see how existing spec/semconv states are mapped to the levels defined here

* The same page lists the status of the different parts of the spec
* The ["Versioning and stability for OpenTelemetry clients"](https://opentelemetry.io/docs/specs/otel/versioning-and-stability/#signal-lifecycle) page has a detailed view on the lifecycle of a signal and which general stability guarantees should be expected by OpenTelemetry clients. Notably, it lacks information about maturity of the Collector. This OTEP could be seen as clashing with the last section of that page, "OpenTelemetry GA". But while that page established a point where both OpenTracing and OpenCensus would be considered deprecated, this OTEP here defines the criteria for calling OpenTelemetry "stable" and making that a requirement for a future graduation. This would also make it clear to end-users which parts of the project they can rely on.
* The OpenTelemetry Collector has its own [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels), which served as inspiration to the ones here.

## Open questions

> * Deliverables such as the Collector Contrib binary and the Java Agent ship with more components than what the SIGs might be comfortable marking as stable. Despite this, there's a desire to mark the deliverables as stable. Would it be sufficient require the SIGs to mark which components are stable?

Reviewer 10: I think it is sufficient if components are clearly marked with any non-stable level.

Author: I believe I covered this in an earlier paragraph now.

> * Should SDKs be required to fully implement the specification before they can be marked as stable?

Author: I split this discussion into its own issue: https://github.com/open-telemetry/opentelemetry-specification/issues/3673

# Definition of maturity levels to be uniformly used by OpenTelemetry SIGs

On 08 Mar 2023, the OpenTelemetry GC and TC held an OpenTelemetry Leadership summit, discussing various topics. One of the themes we discussed was establishing standard rules for describing the maturity of the OpenTelemetry project. This OTEP summarizes what was discussed there and is intended to have the wider community provide feedback.

This OTEP builds on what was previously communicated by the project, especially the [Versioning and stability for OpenTelemetry clients](https://opentelemetry.io/docs/reference/specification/versioning-and-stability).

The Collector's [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels) inspired the maturity levels.

## Motivation

> Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the maturity for components delivered by SIGs in the name of the project. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users using a unified nomenclature.

Reviewer 15: Can we please define what "component" means in this context? We've OTel level components (Collector, instrumentation, etc.) and components within certain projects (Collector components, for instance).

Author: I changed the second occurence of "components" so that it now reads:

> maturity for SIG deliverables in the name of the project

Reviewer 15: > maturity for SIG deliverables (or components) in the name of the project

Something combining both? I think using `components` is still useful if defined properly.

Author: Sounds good, I replaced "or" with "and".

## Explanation

### Maturity levels

> Deliverables of a SIG MUST have a declared maturity level, established by SIG maintainers (SIGs), likely with the input of the code owners. While the main deliverable can have a specific maturity level, individual components might have a different one. Examples:

Reviewer 15: Would it be out of scope to request that maturity level is stated in, say, a `maturity.yaml` file with this and that fields? Similar to what OTel Collector contrib has been doing with components: 

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/metadata.yaml

Reviewer 13: you are referring to https://github.com/open-telemetry/oteps/pull/232/files#r1317870128 - yes, I'd leave that out to our next OTEP or to our tooling.

* the Collector core distribution might declare itself stable and include a receiver that is not stable. In that case, the receiver has to be clearly marked as such
* the Java Agent might be declared stable, while individual instrumentation packages are not

> It is recommended that components should not be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.

Reviewer 15: ```suggestion
Components should not be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.
```

#### Development

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer 15: ```suggestion
Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.
```

#### Alpha

This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The component's interface and configuration options might often change without backward compatibility guarantees. Components at this stage might be dropped at any time without notice.

#### Beta

Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.

> #### Feature freeze

Reviewer 13: I would offer that we name this phase "Stable candidate". "Feature freeze" indicates that we no longer are able to add features, which I don't think is what the text conveys.

Author: Would you mind to expand on that? The text does state that this stage means "feature-complete and ready for broader usage", which does imply that no other features will be added until it's declared stable.

Reviewer 13: I will reframe. I don't think having a stage indicating a "feature freeze" is helpful. I am not a partisan of using feature freeze to declare stability.

Reviewer 10: I think what @atoulme is saying is that the important part of this level is the fact that it is "ready to declared stable", not that no new feature will be added to it. I tend to agree with him. I do not see a good reason to prohibit addition of new features to components that are "ready to be declared stable", provided that it does not disrupt their readiness.

So, perhaps we can adopt the name "RC" as it was originally proposed for this level by @reyang [here](https://github.com/open-telemetry/opentelemetry-specification/issues/3459#issuecomment-1530045325)?

Author: I'm fine renaming this to Release Candidate, but personally,  RC has the notion of a temporary state that should exist for a few days only, while feature freeze implies a longer period. If that's only my impression, I'll change it to RC, as it's not a strong opinion.

Reviewer 10: IMO, RC vs Feature Freeze don't associate with a particular longer or shorter time period.

> The component is feature-complete and ready for broader usage. The component is ready to be declared stable, it might just need to be tested in more production environments before that can happen. Bugs and performance problems are expected to be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

Reviewer 13: Can we get a clear definition of a breaking change as part of this document or pointing to a definition elsewhere?

Author: I had more concrete definitions of what's a breaking change, but it proved too difficult to get a definition that would work for everything: signals, contrib SDKs, agents, collector, ... We delegate now this to each SIG to define.

Reviewer 15: There has be a lowest common denominator. Microsoft defines a breaking change as:

> A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration.

Reviewer 13: That definition works for me. Can we adopt it formally and have each SIG define what their integration surface is (guessing API, config, inputs/outputs)?

Reviewer 13: FWIW I think this can be refined further.
> A breaking change is a change that may require you make changes to your application in order to avoid disruption to your integration, *outside normal change management through communication, deprecation and versioning*.

For example:
Breaking change
> This config option is removed

Non-breaking change
> This config option is deprecated and will be removed in 3 months or 6 releases, whichever comes first. The option is replaced by this other functionality. Please read those migration steps.

Reviewer 13: Here is what I think we can adopt in this OTEP:

```
"A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration."
Microsoft, [LinkedIn breaking change policy](https://learn.microsoft.com/en-us/linkedin/shared/breaking-change-policy)

To understand if a change is to be considered a breaking change, please follow this ruleset:
1. Does the change require users or developers of the component to make themselves changes to avoid disruption of the integration with your component?
   If YES, continue. If NO, the change is considered a non-breaking change.
2. Was the change communicated in advance as per the guidelines of the project, either through release notes, deprecation notice or documentation?
3. Was a mitigation strategy, such as migration steps or workaround offered?
4. Is the change associated with a new version which stipulates the breaking change clearly (using semantic versioning, release notes)?
   If you have replied YES to 2, 3 and 4, the change is considered a non-breaking change. Otherwise, it is considered a breaking change.
```

Author: While that's reasonable for a library or for the collector, that's not an acceptable approach to the protocol: clients are expected to exist for years to come without the expectation that they need to be changed at all. We shouldn't expect new versions of servers to require clients to adopt new versions of our protocol formats in order to communicate with the server, no matter how long in advance we communicated the change.

Author: That said, I do like the original definition: "A breaking change is a change that may require you to make changes to your application in order to avoid disruption to your integration."

Perhaps:

"A breaking change is a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components."

Some examples of breaking changes:
* instrumented applications that need to be changed because the API has been changed
* the SDK starts collecting substantially more telemetry because of a new behavior
* a change to OTLP requires clients to bump their dependency versions in order to communicate with servers on newer versions
* a new semantic convention is added as required

Again, breaking changes are acceptable, just not for stable components.

Reviewer 10: The definition of what is a breaking change is important and it may warrant a separate OTEP. I think we don't necessarily need to have a precise definition of it in order to move forward with this OTEP.

Author: @atoulme, do you think the following is sufficient for now?

> A breaking change is a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components.

Reviewer 13: OK let’s roll

Author: I added this before we start defining the levels. Can you check again?

Reviewer 13: lgtm

#### Stable

The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

#### Deprecated

> Development of this component is halted, and no further support will be provided. No new versions are planned, and the component might be removed from its included distributions. Note that new issues will likely not be worked on except for critical security issues. Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed.

Reviewer 11: > Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed.

I can foresee situations where a component is deprecated and the maintainers do not know which release will happen after six months.  If it is expected that there will be more than two minor releases within less than six months but releases are not on a fixed cadence, what should be communicated?

Author: Good point, I changed it a bit to clarify further.

Author: I'm marking this as resolved, feel free to reopen if you think it's not adequately addressed.

* The same page lists the status of the different parts of the spec
* The ["Versioning and stability for OpenTelemetry clients"](https://opentelemetry.io/docs/specs/otel/versioning-and-stability/#signal-lifecycle) page has a detailed view on the lifecycle of a signal and which general stability guarantees should be expected by OpenTelemetry clients. Notably, it lacks information about maturity of the Collector. This OTEP could be seen as clashing with the last section of that page, "OpenTelemetry GA". But while that page established a point where both OpenTracing and OpenCensus would be considered deprecated, this OTEP here defines the criteria for calling OpenTelemetry "stable" and making that a requirement for a future graduation. This would also make it clear to end-users which parts of the project they can rely on.
* The OpenTelemetry Collector has its own [stability levels](https://github.com/open-telemetry/opentelemetry-collector#stability-levels), which served as inspiration to the ones here.

## Open questions

> * Should SDKs be required to fully implement the specification before they can be marked as stable? See [open-telemetry/opentelemetry-specification#3673](https://github.com/open-telemetry/opentelemetry-specification/issues/3673)

Reviewer 13: I think that given the large scope of the specification and SDKs, the specification should come with a set of functional tests covering portions of the specification, against which SDKs can test and declare themselves stable when they pass all tests.

Author: Can you bring this item to the linked issue? It looks like .NET has some particularities, which would cause it not to pass any compatibility tests based on the SDK.

Reviewer 13: done

> * Should this OTEP define a file name to be adopted by all repositories to declare their deliverables and their maturity levels?

Reviewer 13: I would say no at this point.

Reviewer 15: @atoulme But but but `metadata.yaml`! :)

Reviewer 13: too early to push that everywhere imo.

## Motivation

Quite often, the community is faced with the question of the quality and maturity expectations of its diverse set of components. This OTEP aims to bring clarity by establishing a framework to communicate the maturity for SIG deliverables and components in the name of the project. As the OpenTelemetry project comprises a multitude of SIGs, and each SIG has several components of varying quality, having this framework will help set the right expectations for OpenTelemetry users using a unified nomenclature.

## Explanation

### Maturity levels

Deliverables of a SIG MUST have a declared maturity level, established by SIG maintainers (SIGs), likely with the input of the code owners. While the main deliverable can have a specific maturity level, individual components might have a different one. Examples:

* the Collector core distribution might declare itself stable and include a receiver that is not stable. In that case, the receiver has to be clearly marked as such
* the Java Agent might be declared stable, while individual instrumentation packages are not

> Components should not be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.

Reviewer 16: Nit: SHOULD and MAY are not uppercased consistently in the OTEP

```suggestion
Components SHOULD NOT be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.
```

For the purposes of this document, a breaking change is defined as a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components.

> #### Development

Reviewer 17: Before development, I assume it's possible to have a "Not Implemented" - for example, something defined in the specification that perhaps a client library has not implemented yet. Clearly donating this seems important.

Author: I don't think "not implemented" is a rule for all SIGs. At least for the Collector, the first step is already "in development", which implies that it's not fully implemented.

Reviewer 17: Let's take profiling as an example - if it becomes a new signal with a new procotol or data model, then it would be added to the specification first. Once it's in the specification, it would be added to the collector thus initially being not implemented.

In either case, not implemented should likely be an optional state - client libraries definitely have this today.

Author: It does make sense. Would you mind if I track this as part of a follow-up PR?

https://github.com/open-telemetry/oteps/issues/245

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer: As Antoine said, at least in Go, removing certain components without prior notice causes broken/incorrect documentation. Should this make us drop the 'may be removed without prior notice'?

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer 13: I strongly believe so because this language sent me the wrong way, so I can see it happen again.

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer: cc @jpkrohling PTAL

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Author: I believe that there are two audiences to this statement: maintainers and users. Maintainers should strive to be as friendly to their users as possible, and as such, they should provide prior notice whenever their users experience a change in how the software behaves. This includes marking a Go module as deprecated before removing it entirely without warning.

On the other hand, users should not rely on components that are "under development". If the development for the component hasn't finished yet, it shouldn't be used and users shouldn't be using that in a way that breaks them.

I'd leave this as a MAY, trusting that maintainers will still have their users close to their hearts, while keeping users aware that "under development" components might just disappear.

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer 18: I think the sentence "The component MAY be removed without prior notice." should stay.

MAY does not mean MUST: if there are good reasons to provide notice for a specific language SIG (go), then go maintainers have freedom to provide notice as they see fit.

Removing the sentence means that all languages SIG must provide notice when removing something that is in development status. This is unfair to development, exposing early code to get feedback, with users well aware of risks, should not cause extra burden.

> Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.

Reviewer 13: I can go with what @marcalff said, but we'll need some guidelines in the collector SIG to continue.

#### Alpha

> This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The component's interface and configuration options might often change without backward compatibility guarantees. Components at this stage might be dropped at any time without notice.

Reviewer 17: Why is this the default? Not specifying should default to the earliest status (currently development, but proposed not implemented above)

Author: I _believe_ we discussed this a few months ago (should be part of this PR), but my recollection is that this OTEP will be applied to components that exist today. While we believe that most components (including contrib modules) are beyond the development phase, we don't want to mark them as stable without explicit feedback from their owners.

Author: I'm marking this as resolved but feel free to unresolve, or to open a new issue to address this in a follow-up PR.

#### Beta

> Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.

Reviewer 17: Alpha is the only section that states components might be dropped at any time without notice - is it assumed that is not longer the case once beta is reached?

Author: Given that the interfaces are treated as stable, including configuration and API, dropping a component would likely require a notice. If components can be replaced keeping backward compatibility, no such notice is required.

Reviewer 17: Is this true for beta/RC as well? If so, should be added for clarity.

Author: Yes, beta's description already reads:

> the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible

#### Release Candidate

> The component is feature-complete and ready for broader usage. The component is ready to be declared stable, it might just need to be tested in more production environments before that can happen. Bugs and performance problems are expected to be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are only allowed under special circumstances. Whenever possible, users should be given prior notice of the breaking changes.

Reviewer 17: Feature-complete is not a status in OTel is it? Feature-freeze is a thing. Does RC mean feature-freeze?

Author: Yes, when this stage was proposed, "feature freeze" was the other name suggested. It should also be part of the history of this PR.

Author: Commit where this was changed: 64f2ef7a676b32343d26efea45b3ede2ee37db26

Author: I'm marking this as resolved but feel free to unresolve, or to open a new issue to address this in a follow-up PR.

> #### Stable

Reviewer 17: Everything above this point is "experimental" - a term that's quite common in OTel currently. Would it make sense to have experimental and stable categories with subcomponents?
- Experimental: not implemented, developments, alpha, beta, RC
- Stable: GA, deprecated

Author: This could be a good improvement after this is merged. Given the amount of approvals this one has already, I wouldn't change it right now to add this grouping.

> The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

Reviewer 19: once stable, breaking changes should not be occurring in any circumstances, with or without prior notice, right?

Reviewer 13: No, that's not what the text says. The text says:
> Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
Decomposing this:
> Breaking changes are not expected to happen without prior notice.
Meaning, we will still have breaking changes, but we will organize ourselves around them so they have ample notice. Unless special circumstances which are not defined in this doc.

Reviewer 13: Maybe this sentence should avoid using negatives. A rephrasing might make sense.
```suggestion
The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Changes to configuration options and the component's output are considered breaking changes. Breaking changes are expected to happen with prior notice unless under special circumstances.
```

Reviewer 19: or maybe say - Breaking can occur, as long as semver guidelines are followed? (i.e bump major version.)

Reviewer 13: None of the guidelines here talk about semantic versioning. I guess it's a mix of NIH, trying to steer away from versioning discussions directly, and giving components some flexibility. Add to that Go versioning policies are really strict and will enforce semver in ways that might not feel welcome. You could definitely offer notice by applying the semantic versioning by associating a breaking change with a major version change.

Author: > Breaking changes are expected to happen with prior notice unless under special circumstances.

This flips the meaning of the original text. Breaking changes should be very rare, we should always strive to not have them at all, but we acknowledge that special circumstances might warrant it. The text explicitly attempts to stay away from what's a special circumstance but think of it like a security problem behind a default value.

Reviewer 13: OK, that’s a bit concerning. We need to clarify this.

Author: I see the confusion now:

> without prior notice unless under special circumstances

When taking this as the context, it gives the understanding that prior notice might not be needed under special circumstances. This is wrong. The "unless" refers to the whole statement:

> are not expected to happen without prior notice unless under special circumstances

I tweaked the wording a bit here in an attempt to make it clearer. Can you check again?

Reviewer 13: yes, it's clearer. I was really confused by that sentence. I find it hard to believe we're going to say no more breaking changes once stable though. How do you evolve software without breaking changes (unless special circumstances?) ? Is explicit versioning something we can address maybe down the road to allow folks to give themselves some breathing room?

Author: This OTEP is here to help establish the maturity for deliverables within SIG, and I would argue that a Collector 1.0.0 is one deliverable, while 2.0RC1 is another. They can have different stability levels, while still being part of the same SIG.

Author: @atoulme, @cijothomas, is there something to clarify around this, or can this be marked as resolved?

Reviewer 13: So versioning is key. Thanks Juraci, we're good here then.

> The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

Reviewer 20: Should it be clarified here what can be those special circumstances for breaking change. Or is it left at the discretion of SIG ? One example of special circumstance could be the specification adding a new experimental feature (e.g., Synchronous Instrument) for a already stable API component.

Author: I tried to stay away from defining the examples, as SIGs should be free to determine what's a special circumstance for them.

> The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.

Reviewer 21: I think this question is not directly related to the definition of stability, but it is good opportunity to discuss how we understand _stable_ component which not  support all options form specifications.

I suppose that it is broader project, but lets start with .NET example:
[OTLP exporter for .NET ](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md) is declared stable now. 

It missing some parts of required configuration options like [`OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`](https://github.com/open-telemetry/opentelemetry-specification/blob/89f9bf3b3141626229ac39c9437e1832b38f104f/specification/protocol/exporter.md?plain=1#L7-L17) and [couple others](https://github.com/open-telemetry/opentelemetry-dotnet/issues/4154).

Can it be called stable based on this document? Is it part of maintainer judgement or based on definition it should be never go beyond _Development_ level ("Not all pieces of the component are in place yet")?

---
@cijothomas. FYI

Author: Not exactly the parts you mentioned, but we had a similar discussion here some months ago and I started an issue to tackle this specific aspect: https://github.com/open-telemetry/opentelemetry-specification/issues/3673

I'm marking this as resolved, but if you believe that discussion doesn't relate to your concern, feel free to reopen.

#### Deprecated

> Development of this component is halted, and no further support will be provided. No new versions are planned, and the component might be removed from its included distributions. Note that new issues will likely not be worked on except for critical security issues. Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed, either in terms of a concrete version number or the date of a release, like: "the first release after 2023-08-01".

Reviewer 17: I would propose removing `, and no further support will be provided` - the critical security issue sentence contradicts this.

Author: I think the expectation for all parties should be that support will not be provided. When a security vulnerability appears, maintainers may handle that as a one-off without the requirement to do so.

That said, I removed that statement as I think we can clarify this further in a follow-up PR.

#### Unmaintained

> A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.

Reviewer 17: This feels out of place with everything else. In order to reach any status, it must have had a code owner at some point. If the code owner is not responding, then the component is removed if experimental or marked as deprecated if the component is already stable. This means the existing statuses cover this special circumstance. Instead of this being a maturity level it should be called out separately.

Reviewer 17: I believe this contradicts the point above on "not implemented" if only the collector has this state 😃 

I see the value of this status, especially for developers, but for end-users consuming the release, it seems irrelevant. Even if it's unmaintained it's supported until it's deprecated.

Author: End-users seeing a component they rely on being "unmaintained" might start looking for alternatives or might step in to become code owners themselves (this has happened as well). I agree that developers and other community members are more interested in this than end-users.

Author: I'm marking this as resolved but feel free to unresolve, or to open a new issue to address this in a follow-up PR.

## Trade-offs and mitigations

This OTEP allows SIG maintainers to declare the maturity of the SIG's deliverables without declaring which ones are key for OpenTelemetry. When, and if, this is needed, a new OTEP may be created using the maturity levels as a possible framework.

> ## Prior art and alternatives

Reviewer 17: More:
- https://opentelemetry.io/docs/specs/otel/document-status/
- https://opentelemetry.io/docs/specs/otel/telemetry-stability/ (uses unstable instead of experimental)

Author: Added, thanks!
